text/0000-nested-publish.md

```diff
@@ -0,0 +1,183 @@
+- Feature Name: `nested_publish`
+- Start Date: 2023-06-30
+- RFC PR: [rust-lang/rfcs#3452](https://github.com/rust-lang/rfcs/pull/3452)
+- Rust Issue: ...
+
+# Summary
+[summary]: #summary
+
+Allow Cargo packages to be bundled within other Cargo packages when they are published (not just in unpublished workspaces).
+
+# Motivation
+[motivation]: #motivation
+
+There are a number of reasons why a Rust developer currently may feel the need to create multiple library crates, and therefore multiple Cargo packages (since one package contains at most one library crate). These multiple libraries could be:
+
+* A trait declaration and a corresponding derive macro (which must be defined in a separate proc-macro library).
+* A library that uses a build script that uses another library or binary (e.g. for precomputation or bindings generation).
+* A logically singular library broken into multiple parts to speed up compilation.
+
+Currently, developers must publish these packages separately. This has several disadvantages (see the [Rationale](#rationale-and-alternatives) section for further details):
+
+* Clutters the public view of the registry with packages not intended to be usable on their own, and which may even become obsolete as internal architecture changes.
+* Requires multiple `cargo publish` operations (this could be fixed with bulk publication) and writing public metadata for each package.
+* Can result in semver violations and thus compilation failures, due to the developer not thinking about semver compatibility within the group.
```

Comment:
While some of these indirectly touch on it, one I'd explicitly add is sheer boilerplate.

In working on #3424, one of the things I've noticed is the commentary from people who are looking to further drop boilerplate.  This also came up in a recent blogpost and [HN discussion of it](https://news.ycombinator.com/item?id=36233862)

Reply:
This proposal will still require `Cargo.toml`s for each nested package. What boilerplate do you see removing (besides e.g. explanatory `README.md` files)?

Reply:
Any of the standard manifest fields that crates.io requires.  Granted workspace inheritance helps with those (which will automatically be used in `cargo new` in 1.72) but much nicer if we can just leave them out

Combine that with "cargo script" (if we support `[lib]` packages) and you might not even need manifests (even ones embedded in source)

Reply:
I couldn't find a definitive statement of exactly which fields `crates.io` _requires_ to compare to. <https://doc.rust-lang.org/cargo/reference/publishing.html#before-publishing-a-new-crate> implies it is one, but isn't really (e.g. `homepage` is not mandatory).